controller/pkg/pucontext/pucontext.go

```diff
@@ -116,44 +116,61 @@ func NewPU(contextID string, puInfo *policy.PUInfo, timeout time.Duration) (*PUC
 	return pu, nil
 }
 
-func createACLRules(rules *policy.IPRuleList, dnsrule *policy.DNSRule, ip string) *policy.IPRuleList {
+func createACLRules(rules policy.IPRuleList, dnsrule *policy.DNSRule, ip string) policy.IPRuleList {
 	// ipv6 is not supported
 	if strings.Contains(ip, ":") {
 		return rules
 	}
 
-	rulesAppend := append(*rules, policy.IPRule{
+	rules = append(rules, policy.IPRule{
 		Addresses: []string{ip},
 		Ports:     []string{dnsrule.Port},
 		Protocols: []string{"TCP"},
 		Policy:    dnsrule.Policy,
 	})
 
-	return &rulesAppend
+	return rules
 }
 
 func (p *PUContext) dnsToACLs(dnsList *policy.DNSRuleList, ipcache map[string]bool) {
 
-	rules := new(policy.IPRuleList)
-	for _, dnsrule := range *dnsList {
+	lookupHost := func(dnsrule *policy.DNSRule) error {
+		var rules policy.IPRuleList
+
 		if ips, err := LookupHost(dnsrule.Name); err == nil {
 			for _, ip := range ips {
 				if !ipcache[ip] {
-					rules = createACLRules(rules, &dnsrule, ip)
+					rules = createACLRules(rules, dnsrule, ip)
 					ipcache[ip] = true
 				}
 			}
 
-			if len(*rules) > 0 {
-				if err = p.UpdateApplicationACLs(*rules); err != nil {
+			if len(rules) > 0 {
+				if err = p.UpdateApplicationACLs(rules); err != nil {
 					zap.L().Error("Error in Adding rules", zap.Error(err))
 				}
-				// empty the contents of the rules
-				rules = new(policy.IPRuleList)
 			}
 		} else {
 			zap.L().Warn("Failed to resolve dnsrule", zap.Error(err))
+			return err
 		}
+
+		return nil
+	}
+
+	var errDNSNames policy.DNSRuleList
+
+	for _, dnsrule := range *dnsList {
```

Comment:
I think the construct should be done in parallel with a global timeout (including the retry)

For now afaik if I have 10 dnsrules to resolve that fails I will have to wait 10 times the timeout twice with 500ms in between. 

Also it will avoid to kill baby bears by doing a no lint for an err check ;)